### ISSUE TITLE
[HOLD facebook/hermes release] [$1000] Make emojis larger in other contexts than just single character messages

### ISSUE THREAD

**Issue by jboniface** [Expensify/App#4114]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

This issue covers several related scenarios:

### Action Performed:
Add an emoji with other plain text characters in a message in the composer.

### Expected Result:
The emoji should be larger (19pt) compared to the default size for text (15pt) in the composer. The emoji and text should appear vertically centered with respect to eachother, such that the baseline of the emoji is below the plain text and the middle of the text vertically aligns with the middle of the emoji (like the right-hand-side of [this example](https://github.com/Expensify/App/issues/4114#issuecomment-1078143131).

### Actual Result:
The emoji is the same size as the rest of the text.

### Action Performed:
Send a message with emoji and text.

### Expected Result:
The emoji should be larger (19pt) compared to the default size for text (15pt) in the chat history. The emoji and text should appear vertically centered with respect to eachother, such that the baseline of the emoji is below the plain text and the middle of the text vertically aligns with the middle of the emoji (like the right-hand-side of [this example](https://github.com/Expensify/App/issues/4114#issuecomment-1078143131)

### Actual Result:
The emoji is the same size as the rest of the text.

### Action Performed
Draft a message (of any length) with only emoji.

### Expected Result
The emoji should appear very large (27pt) in the composer.

### Actual Result
The emoji appear very small (15pt).

### Action Performed
Send a message (of any length) with only emoji.

### Expected Result
The emoji should appear very large (27pt) in the report history.

### Actual Result
The emoji only appear large in the report history if there is only one emoji in the message.

### Action Performed
Go to `Settings` -> `Profile` and add emoji(s) to your last name 😎

### Expected Result
The emoji should appear larger (19pt) than the rest of the text (15pt) in the text input. The emoji and text should appear vertically centered with respect to eachother, such that the baseline of the emoji is below the plain text and the middle of the text vertically aligns with the middle of the emoji (like the right-hand-side of [this example](https://github.com/Expensify/App/issues/4114#issuecomment-1078143131)

### Actual Result
The emoji appear the same size as the text.

### Action Performed
1. Go to `Settings` -> `Profile` and add emoji(s) to your last name 😎.
1. Save your profile.
1. Open the `Settings` RHP and look at your name:

### Expected Result
The emoji should appear larger (21pt) than the rest of the text (17pt) in the text input. The emoji and text should appear vertically centered with respect to eachother, such that the baseline of the emoji is below the plain text and the middle of the text vertically aligns with the middle of the emoji (like the right-hand-side of [this example](https://github.com/Expensify/App/issues/4114#issuecomment-1078143131)

### Actual Result
The emoji and text have the same height:

<img width="198" alt="image" src="https://user-images.githubusercontent.com/47436092/160026729-77017294-ae7b-4c0a-a66f-b2310ba7b892.png">

### Action Performed
1. Go to `Settings` -> `Profile` and add emoji(s) to your last name 😎.
1. Save your profile.
1. Open a chat with another user and send them a message. Take a look at your message in the report history.

### Expected Result
The emoji should appear larger (19pt) than the rest of the test (15pt). The emoji and text should appear vertically centered with respect to eachother, such that the baseline of the emoji is below the plain text and the middle of the text vertically aligns with the middle of the emoji (like the right-hand-side of [this example](https://github.com/Expensify/App/issues/4114#issuecomment-1078143131)

### Actual Result
The emoji appears the same size as the rest of the text.

<img width="322" alt="image" src="https://user-images.githubusercontent.com/47436092/160027020-8b45787e-8068-4515-b122-f73c75327d05.png">

## Workaround:
no

## Platform:
all

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @Christinadobrzyn (`AutoAssignerTriage`), see https://stackoverflow.com/c/expensify/questions/4749 for more details.

**Comment by rdjuric** [comment]
@stitesExpensify Do we still want single emojis to be bigger than a emoji in other context (like accompanied by text)? Or we just want to apply the same styling that we apply to single emojis to emojis in all contexts?

**Comment by parasharrajat** [comment]
If this is a pass. I think it is a good time to minimize the regex used for comparison.

**Comment by jboniface** [comment]
this is the original spec:

> The default text size is 15, so when there is text inline with the emoji the emoji will be size 19, and when the emoji is sent by itself it will be size 27.

but I'd like @stitesExpensify to confirm whether we diverged from this plan over the course of design since this feature was on hold for many months.

**Comment by rdjuric** [comment]
Thanks @jboniface. If we end up following the original spec I think a good proposal would be

1. Modify our `ReportActionItemFragment` of type `COMMENT` to check if our `fragment.text` has emojis - but only if it fails our `isSingleEmoji` test.
2. If it has, modify our `Text` to include nested `Text`s containing emojis with a modified style

**Comment by MelvinBot** [comment]
Triggered auto assignment to @ctkochan22 (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by MelvinBot** [comment]
@jboniface, this ***Monthly*** task hasn't been acted upon in 6 weeks; closing.

If you disagree, feel encouraged to reopen it -- but pick your least important issue to close instead.

**Comment by mallenexpensify** [comment]
Why didn't this get worked on?  @ctkochan22 (or another engineer) can this be worked on by an external contributor? If so, can you add the `External` label so we can get this posted? 

I'm going to reopen for now cuz I think it'd be nice to have multiple-emoji messages show larger 
<img width="326" alt="image" src="https://user-images.githubusercontent.com/22508304/157559352-4bd801d4-c13b-4236-a34e-cd3b7a6d026c.png">

**Comment by ctkochan22** [comment]
Yeah this can be worked on by an external contributor. I'm not sure why I didn't add the label, my bad.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @JmillsExpensify (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by JmillsExpensify** [comment]
Oh I love working on this one. Jumping in now.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @parasharrajat (`Exported`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @roryabraham (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by roryabraham** [comment]
I think this issue need to be more specific. I'm really not sure what the requirements are.

**Comment by jboniface** [comment]
Originally I asked @stitesExpensify for specification, but I'm not sure -- maybe we should ask Design? Again, the original spec was:

> The default text size is 15, so when there is text inline with the emoji the emoji will be size 19, and when the emoji is sent by itself it will be size 27.

I'm not sure if that's what was implemented, and the values were just too small or what

**Comment by JmillsExpensify** [comment]
Super fair feedback. @jboniface do you want to take this back to #design?

**Comment by jboniface** [comment]
@JmillsExpensify posted in [slack](https://expensify.slack.com/archives/C03TME82F/p1648055893677569)

**Comment by michelle-thompson** [comment]
Hi, what exactly needs to be fleshed out here? Are we discussing the sizes for emojis in all contexts?

**Comment by jboniface** [comment]
@michelle-thompson yeah, we're trying to clarify what those contexts and sizes should specifically be

**Comment by michelle-thompson** [comment]
If we test the following:
Text - 15pt
Inline emoji - 19pt
Standalone emoji - 27pt

We get something like this, which looks good to me.
![image](https://user-images.githubusercontent.com/43361905/159775897-8462097f-9c3e-4641-9ccd-c03ae473b9a4.png)

The only other suggestion I have is that standalone emojis stay at 27pt, so long as there is no text. Said another way, no max number of emojis before reverting back to 19pt. 

Thoughts?

**Comment by jboniface** [comment]
Yeah, I like that. I think the size limiting is kind of arbitrary when the only thing on the line is emoji.

**Comment by michelle-thompson** [comment]
For sure, just wanted to point it out since it seems like only one emoji is currently allowed before reverting to smaller type.

**Comment by mallenexpensify** [comment]
I like @michelle-thompson 's mockup above.  Since there are only two lines with emoji, do you think there are other examples where text/emoji are mixed and matched (on same and different lines) where they might not look clean? or.. look crowded.  ie. the ❤️ was under Zany's name and there were a line of only large emoji below that

**Comment by michelle-thompson** [comment]
Here's a few more emojis thrown in for good measure! I think it still holds up pretty well.
![image](https://user-images.githubusercontent.com/43361905/159806173-68625962-4531-4f44-9fac-4350f666f082.png)

**Comment by mallenexpensify** [comment]
Looks good, a few of the emoji look like they're 'raised up' a bit, is that 'to be expected' because of the increased size/ratio of emoji to text? 
<img width="934" alt="image" src="https://user-images.githubusercontent.com/22508304/159825936-d4e438d2-3519-499b-9172-ded9bdaff17e.png">

**Comment by roryabraham** [comment]
> a few of the emoji look like they're 'raised up' a bit

Agree, maybe we can fix it by standardizing on a line-height of 19pt to match the larger font of the emoji (even for lines that contain no emoji)? Not sure

**Comment by michelle-thompson** [comment]
Yes, since emojis are a different size they're not gonna be aligned perfectly. You can see in Slack the same also happens:
<img width="208" alt="image" src="https://user-images.githubusercontent.com/43361905/159949857-9621d92a-c40a-4049-976c-1d900b27c0a4.png">

**Comment by michelle-thompson** [comment]
@roryabraham we can certainly try, but I actually don't mind how it looks now! I wouldn't want to increase line height _so_ much that regular text looks oddly spaced out.

**Comment by parasharrajat** [comment]
> Hmmm to my less-trained eyes

Mine too. I think this is achievable via code but the gap between the text lines will be increased.

**Comment by roryabraham** [comment]
> I think this is achievable via code but the gap between the text lines will increase.

Agreed, but that seems worth it to align emoji better. We could try only increasing the line height for lines that contain emoji, but I would prefer to just increase the line height for _all_ lines because:

1. Whether or not a line contains an emoji is hard to calculate (esp when the window resizes, etc...). This would be too hard to get right to be worth even trying imo.
1. We would end up having larger gaps between some lines than others, which would look bad.

**Comment by parasharrajat** [comment]
Yeah, let's see. May be platform will handle it automatically. Otherwise, I agree with your opinion.

**Comment by michelle-thompson** [comment]
Yeah, I agree that this proposed line height apply to all text, not just lines of text with an emoji. Also, definitely open to seeing how center-aligned emojis will look!

**Comment by roryabraham** [comment]
Okay, I took a stab at clarifying the requirements in the OP. Feel free to review, but I feel much better about moving forward with this issue now :)

**Comment by roryabraham** [comment]
@JmillsExpensify if the updated OP looks good to you can you please update the Upwork job to match the OP and bump it to $500?

**Comment by JmillsExpensify** [comment]
Awesome, great convo! Updated Upwork on both counts.

**Comment by JmillsExpensify** [comment]
Approaching a week and still no takers. Given that time preceding our most recent convos, I'm going to raise the price to $1,000 and update the issue title to encourage more contributors to jump in.

**Comment by JmillsExpensify** [comment]
It's been a week, so increasing the payout for the issue once more to $2,000

**Comment by JmillsExpensify** [comment]
We're on the even of 7 days here, and this would still be great to do, so I'm going to bump once more to $4,000 and then we can decide next week if it's worth going any higher.

**Comment by JmillsExpensify** [comment]
@roryabraham I'm curious if you have any theories on why we're still not getting any proposals yet. Is it the difficulty of the issue?

**Comment by eVoloshchak** [comment]
**Scenario 4:**
1. Add the following function to [EmojiUtils](https://github.com/Expensify/App/blob/main/src/libs/EmojiUtils.js):
```javascript
function containsOnlyEmojis(message) {
    const match = message.match(CONST.REGEX.EMOJIS);
    return match && match.length === [...message].length;
}
```
2. Add this line in [ReportActionItemFragment](https://github.com/Expensify/App/blob/8f654504b989bf78f141b72504e48e5993a1ab7f/src/pages/home/report/ReportActionItemFragment.js#L99) (we will need this variable for the Scenario 2, hence i'm calling it here to avoid calling it twice)
```javascript
const containsOnlyEmojis = EmojiUtils.containsOnlyEmojis(props.fragment.text);
```
3. Replace [this](https://github.com/Expensify/App/blob/8f654504b989bf78f141b72504e48e5993a1ab7f/src/pages/home/report/ReportActionItemFragment.js#L109) line with the
```javascript
style={[containsOnlyEmojis ? styles.singleEmojiText : undefined]}
```
Result:
![Screenshot from 2022-04-19 17-25-31](https://user-images.githubusercontent.com/9059945/164026798-3c81f500-20ed-4bc7-85c0-431762779ba2.png)

**Scenario 2:**
1. Add the following function to [EmojiUtils](https://github.com/Expensify/App/blob/main/src/libs/EmojiUtils.js):
```javascript
function isEmoji(char) {
    return Boolean(char.match(CONST.REGEX.EMOJIS));
}
```
2. Add the necessary [variables](https://github.com/Expensify/App/blob/main/src/styles/variables.js) 
```javascript
fontSizeEmoji: 19,
fontSizeEmojiHeight: 21,
```
3. And [styles](https://github.com/Expensify/App/blob/main/src/styles/styles.js) 
```javascript
emojiMessageText: {
    fontSize: variables.fontSizeEmoji,
    lineHeight: variables.fontSizeEmojiHeight,
},
```
4. In `ReportActionItemFragment.js` replace [this](https://github.com/Expensify/App/blob/add4755bbfc91ccb644462cff02c4746714e8f24/src/pages/home/report/ReportActionItemFragment.js#L111) line with
```javascript
{containsOnlyEmojis 
    ? Str.htmlDecode(props.fragment.text)
    : [...Str.htmlDecode(props.fragment.text)].map((char => 
        <Text style={{textAlignVertical: 'center', lineHeight: styles.emojiMessageText.lineHeight, fontSize: EmojiUtils.isEmoji(char) ? styles.emojiMessageText.fontSize : undefined}}>
            {char}
        </Text>
    ))
}
```
Result:
![Screenshot from 2022-04-19 17-33-16](https://user-images.githubusercontent.com/9059945/164028510-0e86b223-26dd-4294-99ac-583a08a01881.png)

**Scenario 6:**
1. Add the necessary [variables](https://github.com/Expensify/App/blob/main/src/styles/variables.js) 
```javascript
fontSizeEmojiProfile: 21,
fontSizeEmojiProfileHeight: 23,
```
2. Add `lineHeight` to [displayName](https://github.com/Expensify/App/blob/add4755bbfc91ccb644462cff02c4746714e8f24/src/styles/styles.js#L1863) styles
```javascript
lineHeight: variables.fontSizeEmojiProfileHeight,
```
3. Add the `renderUserName` function to [InitialSettingsPage](https://github.com/Expensify/App/blob/add4755bbfc91ccb644462cff02c4746714e8f24/src/pages/settings/InitialSettingsPage.js#L151)
```javascript
const renderUserName = (name) => [...name].map((char => 
    <Text style={{textAlignVertical: 'center', fontSize: EmojiUtils.isEmoji(char) ? styles.emojiProfileText.fontSize : undefined}}>
        {char}
    </Text>
));
```
4. And call it on [this](https://github.com/Expensify/App/blob/add4755bbfc91ccb644462cff02c4746714e8f24/src/pages/settings/InitialSettingsPage.js#L173) line 
```javascript
? renderUserName(props.myPersonalDetails.displayName)
```
Result:
![Screenshot from 2022-04-19 17-40-26](https://user-images.githubusercontent.com/9059945/164030025-4f28634d-3ccd-47cb-a27a-20df6e1cc7cb.png)

**Scenario 7:**
In `ReportActionItemFragment.js` replace this [line](https://github.com/Expensify/App/blob/add4755bbfc91ccb644462cff02c4746714e8f24/src/pages/home/report/ReportActionItemFragment.js#L130) with 
```javascript
{[...Str.htmlDecode(props.fragment.text)].map((char => 
    <Text style={{textAlignVertical: 'center', fontSize: EmojiUtils.isEmoji(char) ? styles.emojiMessageText.fontSize : undefined}}>
        {char}
    </Text>
))}
```
Result:
![Screenshot from 2022-04-19 17-44-09](https://user-images.githubusercontent.com/9059945/164030830-b169db8b-26c3-4249-84a2-8beb1bf1b82a.png)

As for the scenarios 1, 3 and 5, they require conditional rendering fonts of different sizes inside the `TextInput`. I don't think this is possible, I've tried researching it and asking my colleagues, but no luck. I'm going to research it a bit more and apdate my proposal in case I find something, but if you have some knowledge about it - I would highly appreciate it if you share. Maybe ther'se an app that has this behaviour (i.e. different font sizes insed the `input`)?

**Comment by parasharrajat** [comment]
I will review the proposal in details shortly but starting thoughts are: 

> As for the scenarios 1, 3 and 5, they require conditional rendering fonts of different sizes inside the TextInput. I don't think this is possible, I've tried researching it and asking my colleagues, but no luck. I'm going to research it a bit more and apdate my proposal in case I find something, but if you have some knowledge about it - I would highly appreciate it if you share. Maybe there's an app that has this behavior (i.e. different font sizes inside the input)?

I agree with this as well. I can't imagine a solution where we can have multiple font sizes on textInput. 

But I think we can use multiline TextInput on the Web to render HTML. No idea about the native.

**Comment by parasharrajat** [comment]
There is no solution proposed for scenarios 1, 3, and 5. And I don't know if that is possible for cross-platform. 

@JmillsExpensify Let's discuss these requirements first before I pick the proposal.

**Comment by parasharrajat** [comment]
Asked here https://expensify.slack.com/archives/C01GTK53T8Q/p1651146286301709

**Comment by JmillsExpensify** [comment]
Thanks! Coming back from out-of-office and I just commented on your post.

**Comment by JmillsExpensify** [comment]
Based on the discussion in the linked Slack thread in #expensify-open-source, I'm going to close this issue so we can prioritize alternative roadmap priorities. Thank you everyone!

**Comment by JmillsExpensify** [comment]
@stitesExpensify I wanted to circle back on this issue now we're prioritizing emojis again. We actually never did anything here and ended up closing the issue with no action. It might be good to resurrect this issue/convo and see if we can/should tackle it with a fresh set of eyes.

**Comment by stitesExpensify** [comment]
Agreed, I think it would be a great addition! Do you think we should re-open this one, or create a new one? I think re-opening this one makes sense considering how detailed the OP is

**Comment by JmillsExpensify** [comment]
Nice, agreed. I just re-opened. So next up, are we going to keep this external, such that we're open again for proposals?

**Comment by JmillsExpensify** [comment]
Related, I thought it fair to update this issue back to the beginning at this point.

**Comment by stitesExpensify** [comment]
> are we going to keep this external, such that we're open again for proposals

Yes, I think that we should start taking proposals again!

**Comment by JmillsExpensify** [comment]
Nice, I'll float this one in open-source. It's got `Help Wanted` on it but it's old so might not be on people's radar.

**Comment by josemak25** [comment]
Hi I am from Callstack and would like to take this and be assigned to this task.

**Comment by MelvinBot** [comment]
📣 @josemak25 You have been assigned to this job by @pecanoro!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by mountiny** [comment]
The issue was not linked properly.

**Comment by JmillsExpensify** [comment]
Also wow that PR happened very quickly!

**Comment by bernhardoj** [comment]
@parasharrajat does this issue also handle emoji inside markdown?

**Comment by bernhardoj** [comment]
I mean when we send a message like this:

``` 
italic -> _🥳_
bold -> *🥳*
etc..
```

Currently, all emoji inside markdown has a small size.

**Comment by stitesExpensify** [comment]
Sorry if I'm confused here, but Emojis should not be affected by markdown. If you were to type `*😄*` that would not be a single character message

**Comment by bernhardoj** [comment]
Thanks for the confirmation! So, this won't be an [issue](https://github.com/Expensify/App/issues/15424#issuecomment-1448405819) then, right?

**Comment by stitesExpensify** [comment]
Correct!

**Comment by ArekChr** [comment]
I'm taking over @josemak25 tasks, could you assign me?

**Comment by MelvinBot** [comment]
📣 @ArekChr You have been assigned to this job by @mountiny!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by MelvinBot** [comment]
This issue has not been updated in over ***15*** days. @JmillsExpensify, @parasharrajat, @ArekChr, @roryabraham eroding to ***Monthly*** issue.

P.S. Is everyone reading this sure this is really a near-term priority? Be brave: if you disagree, go ahead and close it out. If someone disagrees, they'll reopen it, and if they don't: one less thing to do!

**Comment by roryabraham** [comment]
@fabioh8010 is working on updating Hermes to support the `-d` flag in regexes. Most recent updates are [here](https://github.com/Expensify/App/pull/15611#issuecomment-1485884839)

**Comment by JmillsExpensify** [comment]
Latest updates in the linked PR.

**Comment by JmillsExpensify** [comment]
I lost track where we're at with this issue? Can someone help clarify?

**Comment by fabioh8010** [comment]
Hi @JmillsExpensify , in order to continue development we need to implement a new feature in Hermes, [which is done and is currently in review](https://github.com/facebook/hermes/pull/968). It got another review and I need to do some changes, I'll start addressing today.

**Comment by fabioh8010** [comment]
**Update:** Addressed the requested changes on May 14th, waiting for another review at the moment.

**Comment by JmillsExpensify** [comment]
Thanks for the updates!

**Comment by stitesExpensify** [comment]
Looks like we're still waiting for another review on the hermes PR.

**Comment by fabioh8010** [comment]
**Update:** Asked the reviewer if he had a chance to take a look at the PR.

**Comment by fabioh8010** [comment]
**Update**: Addressed more requested changes on the PR, waiting for another review now.

**Comment by stitesExpensify** [comment]
The PR is on hold for now while we're updating Hermes

**Comment by fabioh8010** [comment]
**Update:** Still waiting for Meta's team decision about how to merge the Hermes PR.

**Comment by roryabraham** [comment]
Still in the pipeline, awaiting https://github.com/facebook/hermes/pull/968 to land in RN 0.73

**Comment by roryabraham** [comment]
Milestone: https://github.com/facebook/hermes/pull/968 was merged 2 weeks ago. Still not included in a Hermes release

**Comment by JmillsExpensify** [comment]
Still not done.

**Comment by JmillsExpensify** [comment]
Same

**Comment by roryabraham** [comment]
Whoa, this might actually be the oldest active issue in NewDot, and at long last it's coming off HOLD. Going to dust this off and bring it up in slack...

**Comment by roryabraham** [comment]
Confirmed, it's the oldest 😉 

![image](https://github.com/Expensify/App/assets/47436092/08f454cf-2518-424e-8ae9-506d99a080a3)

**Comment by MrRefactor** [comment]
Hi! Its Kuba from Callstack team and I would like to take this issue :D

**Comment by hayata-suenaga** [comment]
I assigned you @MrRefactor

**Comment by MrRefactor** [comment]
Thanks @hayata-suenaga !

**Comment by hayata-suenaga** [comment]
by the way, I have no idea why I got notification for this issue, but I went ahead and assigned Jakub 🙇

**Comment by roryabraham** [comment]
@MrRefactor any estimate on when you'll have a PR ready for review?

**Comment by MrRefactor** [comment]
@roryabraham  Looking at both issues, Im aiming to create PR mid next week

**Comment by MrRefactor** [comment]
Still working on the issue, dont have any estimations atm, will try to create PR soon

**Comment by roryabraham** [comment]
Looking forward to reviewing!

**Comment by kbieganowski** [comment]
Hi. I'm Kamil from Callstack - expert contributor group - I would like to take over this task from Kuba

**Comment by MrRefactor** [comment]
> Hi. I'm Kamil from Callstack - expert contributor group - I would like to take over this task from Kuba

I didnt have enough time last week to work on this issue so Kamil is taking this over 

@roryabraham

**Comment by roryabraham** [comment]
we're discussing reassigning this [here](https://expensify.slack.com/archives/C03UK30EA1Z/p1709556279070389?thread_ts=1709081216.812279&cid=C03UK30EA1Z)

**Comment by kbieganowski** [comment]
Update:
I was able to make some progress with this ticket and opened wip PR that [can be found here](https://github.com/Expensify/App/pull/37980).
Issue I'm facing at the moment is that on iOS even though we are extracting emojis (with surrogate pairs) from the string, not every (surrogate pair) emoji is responding to applied styles - for example in iOS screen from the PR we have 2 surrogate pair emojis (😶‍🌫️ and 🇦🇩), if we change value of fontSizeEmojisWithinText to noticeably bigger value, flag emoji will be bigger as expected but size of 😶‍🌫️ emoji stays the same which is not an issue on android and web - both surrogate pairs, both recognized as emojis (red background) but different outcomes.

After this issue is resolved it should be smooth sailing to update remaining places mentioned in ticket description.

Regarding emojis size inside of the composer, I'm figuring out what would be the best and non-hacky way to approach this, I have few ideas in mind and will investigate which one would be best candidate
cc: @roryabraham

**Comment by roryabraham** [comment]
> Regarding emojis size inside of the composer

Please post in #expensify-open-source and tag @tomekzaw, @thienlnam, and myself

**Comment by kbieganowski** [comment]
Update:
I've resolved issue that I've mentioned in last update - solutions suggested in previous PRs had problems with flags, skin tone modifiers and some surrogate pairs variations, with updated `splitTextWithEmojis` we are handling all of that so no emoji variation should be an issue from now on which was our biggest problem.

At the moment I'm solving iOS and Android layout issues where on iOS some emojis are pushed to new line and on Android height of emojis is not considered in emojis-only messages (both issues can be seen in screenshots of [WIP PR](https://github.com/Expensify/App/pull/37980/files#diff-1692fc6ebb559e5fd701e993d1e71ebee2b6ed378c84d2a04654b3ff3ff20c50R567))
cc: @roryabraham

**Comment by tomekzaw** [comment]
FYI I've added support for customizing `emoji.fontSize` in https://github.com/Expensify/react-native-live-markdown/issues/232 but we're waiting for `<emoji>` tag to re-land in `expensify-common/ExpensiMark`.

**Comment by roryabraham** [comment]
looks like the [PR](https://github.com/Expensify/App/pull/37980) is still WIP.

@kbieganowski any status update?

> we're waiting for <emoji> tag to re-land in expensify-common/ExpensiMark.

does anyone know where this is being tracked?

**Comment by roryabraham** [comment]
@kbieganowski can you clarify if it's possible for you to make progress on this issue before `emoji.fontSize` is supported in react-native-live-markdown?

**Comment by kbieganowski** [comment]
@roryabraham - it really depends on what is eta for `emoji.fontSize` support to land. Status of this ticket (putting aside emojis size in inputs since it is handled by separate PR) is that logic for splitting emojis and text (including all emoji variations) is done, components to display larger emojis in username and comments are updated and working as expected on web and desktop. Right now I'm resolving layout issues on native android and ios for comments with updated emojis. I'm hoping to have it ready for review by the end of week

**Comment by tomekzaw** [comment]
@kbieganowski `emoji.fontSize` support in Live Markdown is ready, just waiting for re-landing `<emoji>` tag in ExpensiMark and we're ready to merge/release

**Comment by roryabraham** [comment]
`<emoji>` tags have been re-added in Expensimark. So it seems like with a good push we should be able to finish this week?

**Comment by tomekzaw** [comment]
* Bumped expensify-common in https://github.com/Expensify/react-native-live-markdown/pull/233.
* Finished the PR so that `emoji.fontSize` style in Live Markdown now works as expected.
* Asked @BartoszGrajdek for a quick review.
* We should be able to merge on Monday.

**Comment by kbieganowski** [comment]
Update:
Scenario 1 ✅
Scenario 2 ✅
Scenario 4 ✅
Scenario 7 ✅

Scenarios 3,5 and 7: This is mostly done but at the moment I'm resolving issues on both ios and android where after changing input to `RNMarkdownTextInput` and passing value prop, on android cursor is jumping to the start after writing 1st character or adding emoji and on ios emoji picker (or pasting emojis) stopped working. After that issue is resolved, last thing to be done is adjusting paddings/height of composer on mobile so larger emojis are centered and not cut from the top.

Note: I had to adjust line height in comments a little bit so enhanced emojis are not overlapping or cut from the top in some cases - this was done for comments both with and without emojis so the UI can stay consistent. The difference is marginal but eliminates issues with taller emojis (I will put that info in PR description and tag someone from design team when it's ready for review).

Scenarios marked as done can be checked in web and desktop recordings in draft PR (I skipped native recordings for now since there are obvious issues described above).
cc: @roryabraham

**Comment by kbieganowski** [comment]
Update:
Fixed issue of cut emojis when we have emojis only in composer (on native ios and android) as well as some other minor UI issues.
Things left to be done:
1. After discussion with @tomekzaw and @Skalakid new issue was opened in rn live markdown repo (https://github.com/Expensify/react-native-live-markdown/issues/293). I will bump the version once this is resolved
2. Scenario 5 - changing current input to `RNMarkdownTextInput` breaks input styles and functionality, I will investigate this first thing monday. After this 2 things are resolved we should be good for review🤞
Note:
Scenario 6 - I think this AC is no longer valid (text font size for user's name and last name in LHP is set to 22px now)

**Comment by roryabraham** [comment]
thanks for pushing this forward!

> Scenario 6 - I think this AC is no longer valid (text font size for user's name and last name in LHP is set to 22px now)

Scenario 6 is still valid:

<img width="201" alt="image" src="https://github.com/Expensify/App/assets/47436092/4dd6ecd2-3912-4531-8bb3-094db944ec9a">

**Comment by kbieganowski** [comment]
Hello @roryabraham, a bit of an update on this one:

I've updated user's last name input in settings page so emojis are enhanced as well as emoji fontSize in settings LHP (according to your last comment about scenario 6, please note that AC says that emoji fontSize should be enhanced to 21px while current fontSize of this section is 22px so I've changed fontSize for emojis to 25px, let me know if that works for you).

I've opened new draft PR for this ticket https://github.com/Expensify/App/pull/40692 (amount of conflicts in previous PR was not worth wasting the time to resolve them commit by commit)

It seems like there are 2 last things that block us from moving this ticket to internal review:

1. iOS issue mentioned in previous update (PR to rn live markdown with fix is opened and awaiting review, this will be quick change of version in package.json and that's all)

2. There is an issue with how comments are rendered in chat (not related to my changes but I would guess this will be most likely blocked by QA team) - there is some kind of race condition while rendering comments that causes emojis regex check to return different outcomes on pretty much each render, change like adding `console.log('1234')` or some empty loop is causing different render outcome. I've added red bg to emojis within text so it is easier to catch it in for example recording from chrome where 1st comment from the top applies expected styles but emojis in 2nd comment are not styled (I would get different outcome if I would add some random `console.log()` or loop before render). I've also noticed that sometimes emojis in some comments are styled on 1st render and then not after comments are re-rendered...

Please let me know if we want to create separate ticket for this issue and move this PR forward or investigate the issue in scope of this ticket/PR?

**Comment by roryabraham** [comment]
I'm not sure I have a good understanding of the issue you're describing in `2` unfortunately. Can you please report the bug in slack using the template of:

### Actions Taken (step-by-step)

### Expected Result

### Actual Result

then we can discuss what to do once we have a better understanding.

**Comment by kbieganowski** [comment]
Update:
After long conversation with @fabioh8010 in [this PR](https://github.com/Expensify/App/pull/40692) as well as internal discussions we decided to went with different implementation of text and emojis splitting logic based on solution previously suggested by @roryabraham (with some small tweaks) - it should be bit more performant and easier to work on in the future.
 
As mentioned by Fabio, there are few other places that were previously using emoji regex with global flag and I'll need to validate if there are no issues after regex changes (and provide small fixes if needed), important note is that after changes to logic and minor UI updates we have same outcomes as before in every place so no progress that was already made is lost.

Another positive outcome of removing global flag from regex is that it most likely caused unexpected behaviours in other places, this change will eliminate that.

I'm leaving this PR as a draft for now since we have 3 ongoing topics:
1. Validation I mentioned above
2. [RN live markdown PR](https://github.com/Expensify/react-native-live-markdown/pull/318)
3. Discussion regarding issues with comments rendering [in slack](https://callstack-hq.slack.com/archives/C03UK30EA1Z/p1713968506041209?thread_ts=1713968151.471429&cid=C03UK30EA1Z)

Note: I'll be OoO next week (will be back on May 6th)

**Comment by kbieganowski** [comment]
Update:
I've applied all the suggestions in draft PR and we are good on that end. 2 issues blocking this ticket at the minute are the things mentioned previously:

1. Opened PR to RN live markdown
2. Discussion we started previously here and on slack regarding how comments are rendered - I've investigated this issue a little bit but wasn't able to find the root cause. Please let me know if we want to investigate it more in scope of that ticket/PR or create separate issue so this ticket is no longer blocked

Also, can we add `Design` badge to issue and [PR](https://github.com/Expensify/App/pull/40692) (I don't have permissions to do so)? - As I mentioned [in this comment](https://github.com/Expensify/App/issues/4114#issuecomment-2039827060), line height in the comments was increased a little bit so larger emoji and text in line above it are not overlapping

cc @roryabraham

**Comment by roryabraham** [comment]
The `Design` team will be automatically brought in for a review when you take the PR out of draft since they are code owners of the `styles` directory.

**Comment by roryabraham** [comment]
> Discussion we started previously here and on slack regarding how comments are rendered - I've investigated this issue a little bit but wasn't able to find the root cause. Please let me know if we want to investigate it more in scope of that ticket/PR or create separate issue so this ticket is no longer blocked

Let's create a separate GitHub issue for this and not block this issue on it

**Comment by VickyStash** [comment]
Hi, I'm Viktoryia from Callstack - expert contributor group - and I would like to take over this issue 🙂

**Comment by JmillsExpensify** [comment]
Assigning you to the issue!

**Comment by VickyStash** [comment]
Updates:
[The PR](https://github.com/Expensify/App/pull/40692) is waiting for the C+ review

**Comment by VickyStash** [comment]
Note: I'll be OOO Jul 1 - Jul 7 🌴

**Comment by roryabraham** [comment]
PR merged!!!

**Comment by VickyStash** [comment]
The [PR](https://github.com/Expensify/App/pull/40692) got [reverted](https://github.com/Expensify/App/pull/46463).
I'm going to look through all of the blockers within the app and react-native-live-markdown lib before opening a new PR.

**Comment by VickyStash** [comment]
**Updates:** I have not made much progress here as of today since I was mostly focused on another issue. But I should be able to dive in more starting tomorrow.

**Comment by VickyStash** [comment]
**Updates:** I've started to work on the new PR, created one more issue related to react-native-live-markdown: https://github.com/Expensify/react-native-live-markdown/issues/448

**Comment by VickyStash** [comment]
**Updates:** today I've been focusing on the report screen display - display names and messages.
I'm trying to keep the initial `lineHeight`, though on ios due to rn issues with emojis alignment within text it's not so easy. That's what I have as for now:

| Before  | After |
| ------------- | ------------- |
| ![Simulator Screenshot - iPhone 15 - 2024-08-02 at 17 12 27](https://github.com/user-attachments/assets/f77b07da-d14f-4415-8e11-e706563a26e2) | ![Simulator Screenshot - iPhone 15 - 2024-08-02 at 17 09 03](https://github.com/user-attachments/assets/71f4bb81-ffb1-48d5-9aa8-8d4a6f2e5f03) |
| <img width="1006" alt="before_web" src="https://github.com/user-attachments/assets/8d02cb38-7955-44b8-90f4-e97b45e53602">  | <img width="1008" alt="after_web" src="https://github.com/user-attachments/assets/ce693c1b-5054-4573-b80f-5716a03de46c"> |

**Comment by shawnborton** [comment]
It still feels like we aren't quite getting even line spacing though. Do we need to actually make the emojis smaller? 

Basically the gaps in the lines here do not look even:
![CleanShot 2024-08-05 at 09 28 44@2x](https://github.com/user-attachments/assets/11ede3d3-ca14-4afc-82fc-0cc8f7d54214)

**Comment by VickyStash** [comment]
@shawnborton Please, take a look at some options:

| fontSize : 15, lineHeight: 20  | fontSize : 17, lineHeight: 20| fontSize : 17, lineHeight: 22 | fontSize : 19, lineHeight: 22 |
| ------------- | ------------- | ------------- | ------------- |
| <img width="349" alt="1_FS15" src="https://github.com/user-attachments/assets/2f23ac52-07c2-474a-bd57-cbde2a36be80">  | <img width="349" alt="2_FS17" src="https://github.com/user-attachments/assets/d8f4813b-d6ef-4c5d-bf17-454190e86355">  | <img width="347" alt="3_FS17_22" src="https://github.com/user-attachments/assets/6692da33-c948-460c-9a03-bfd4d68afd23"> | <img width="349" alt="4_FS19_22" src="https://github.com/user-attachments/assets/9b57f3a2-75e3-4307-a6f0-b1d8b14773c2">  |

**Comment by shawnborton** [comment]
Sorry if this wasn't clear before, but we definitely do not want to change the line height. We want to keep our standard font line height of 20px here.

So it seems like the only way this works is if we don't actually change the font size of the emojis at all... which might defeat the purpose of this PR, but personally I think it makes the most sense unless we can make it so that larger emojis don't mess up the line height alignment. Any ideas there?

**Comment by VickyStash** [comment]
@shawnborton Small note, even on the current main/release, there is inconsistent line spacing on ios:
![image](https://github.com/user-attachments/assets/d92b3f32-eb9d-4293-bd1e-6a886a4758ad)

Also, the emojis are cut off in some places through the app.

I was able to fix it by applying `System` font size to emojis only:
![Simulator Screenshot - iPhone 15 - 2024-08-06 at 12 28 43](https://github.com/user-attachments/assets/8034738e-36ce-4ed4-9e20-216227e7f18b)

So at least we can fix issues with current emoji size display.
But I'll do a couple more tries to fit the large emoji as well.

**Comment by shawnborton** [comment]
Sounds good, thanks for continuing to dig into this one!

**Comment by JmillsExpensify** [comment]
Removed payment from the title since we're still working on this one.

**Comment by VickyStash** [comment]
@shawnborton What do you think about increasing emojis font size on ios to 17 and keep the same line height:
![Simulator Screenshot - iPhone 15 - 2024-08-09 at 18 55 56](https://github.com/user-attachments/assets/802408ac-42f5-402b-8510-e1bb53e5937c)

**Comment by shawnborton** [comment]
Dang, that feels pretty good to me! What do you think? Also cc @Expensify/design for thoughts too. 

That feels like the best happy medium we've found so far in terms of making the emoji size slightly bigger but retaining the same line height as the other text.

**Comment by dannymcclain** [comment]
>That feels like the best happy medium we've found so far

Agree—I think this is looking pretty good!

**Comment by dubielzyk-expensify** [comment]
Hmm. Does this mean the font size in chat will be different than everywhere else in the app? 16 vs 17?

**Comment by VickyStash** [comment]
> Hmm. Does this mean the font size in chat will be different than everywhere else in the app? 16px vs 17px?

@dubielzyk-expensify We change only emojis font size, that's the intention of this issue. The text size is left the same - in most of the cases 15px.

**Comment by shawnborton** [comment]
Yup, that sounds right to me. The idea being that emojis should appear slightly larger, otherwise when they are the same font size as the text, they can feel a bit small.

**Comment by VickyStash** [comment]
**Updates:** today I was focused on the web display of the composer and the report screen, that's what I have for now:

| Before  | After |
| ------------- | ------------- |
|  ![before](https://github.com/user-attachments/assets/dd07bebb-ce99-4f1b-806f-715344adb326) | ![after](https://github.com/user-attachments/assets/0276ff1d-51ec-4863-9178-1bf9a299ee4a)  |

On the web, the 17 font size for emojis also fits better, especially when it comes to text selection.

**TODOs:**
- implement scenario 5 and 6
- recheck all of the reported issues during the first PR merge

**Questions:**
This issue mentiones increasing emojis size to 27px font size if the message consists of only emojis. But it looks like this already was implemented, with the font size 30px though, should it be changed?

**Comment by shawnborton** [comment]
Can you share what the 30px emojis look like? Maybe we can just keep them as-is.

**Comment by dubielzyk-expensify** [comment]
> @dubielzyk-expensify We change only emojis font size, that's the intention of this issue. The text size is left the same - in most of the cases 15px.

I see. Thanks for clarifying 😄

**Comment by VickyStash** [comment]
> Can you share what the 30px emojis look like? Maybe we can just keep them as-is.

@shawnborton That's how it looks right now on prod:
![image](https://github.com/user-attachments/assets/dd55875e-ecb3-4434-82fe-cd70cc293d1c)

**Comment by shawnborton** [comment]
Okay cool, I think the size of that looks fine. But I think we should consider changing the `line-height: 35px` to something like 20px, so we can make it feel like the composer doesn't shift in size even when the larger emoji is present, like so:
![CleanShot 2024-08-13 at 06 29 53@2x](https://github.com/user-attachments/assets/4fae339a-e70c-4729-9060-a5c6210bdd32)

Let me know if that makes sense.

**Comment by VickyStash** [comment]
@shawnborton With 20px it has jumping selection line 
![image](https://github.com/user-attachments/assets/8ce26c29-5bb6-4a8c-bf29-fab88ecf7216)

At this point, 27px font size looks better
![image](https://github.com/user-attachments/assets/24e7db47-1326-48cd-881f-af6ef0838e35)

**Comment by shawnborton** [comment]
Hmm but it sitll looks like the composer is larger/taller than it needs to be. Check my screenshot immediately above and let me know what you think.

**Comment by VickyStash** [comment]
@shawnborton Okay, I got what you did and did the same on my side. The composer's height looks okay, but if there are several lines of the emojis, they are overlapped.
![image](https://github.com/user-attachments/assets/4d66f761-45e0-4228-9dc0-c72707b237a0)
![image](https://github.com/user-attachments/assets/22d4cd67-9449-4530-ad05-38b8d4f368f6)

But, I've tried to apply increased lineHeight to multiline emojis, and that's how it can look

https://github.com/user-attachments/assets/28b85475-f487-429d-9d81-28b167f717f0

**Comment by shawnborton** [comment]
Cool, thanks for exploring that. I feel like we should mimic other apps where if there are more than 3-4 emojis, we go back to the regular text size. I believe Slack does something like this right?

**Comment by VickyStash** [comment]
> Cool, thanks for exploring that. I feel like we should mimic other apps where if there are more than 3-4 emojis, we go back to the regular text size. I believe Slack does something like this right?

@shawnborton Slack goes back to the regular text size only after 23rd emoji 😄
But I think this is a little bit out of scope, we can add it as a follow-up later.

**Comment by shawnborton** [comment]
Hmm but if we add it as a follow up later, that means we're going to run into the emoji overlap issue if we make it so that the composer doesn't need to grow when just a few large emojis are present. So I would prefer to just solve this all at once if we can.

**Comment by VickyStash** [comment]
@shawnborton Did you take a look at the video from my comment above? The emojis doesn't increase the composer size when they are in one line and when there are several lines they are not overlapped.

> But, I've tried to apply increased lineHeight to multiline emojis, and that's how it can look

https://github.com/user-attachments/assets/28b85475-f487-429d-9d81-28b167f717f0

But if it doesn't work to you I'm okay to update the logic with amount of emojis affecting the emojis size. Let's just mention it under Scenario 3 and 4 in the issue.

**Comment by shawnborton** [comment]
Ah okay, well then that would work for me but I'm curious how we would handle textwrapping when there is say 30 emojis in a row on various screen widths?

**Comment by VickyStash** [comment]
@shawnborton It looks this way (can't do android, cause dev seems to be down):

<img width="1276" alt="Monosnap (2) New Expensify 2024-08-13 15-43-13" src="https://github.com/user-attachments/assets/ee685bea-fede-43c2-9c34-650d501f157f">

![Simulator Screenshot - iPhone 15 - 2024-08-13 at 15 43 00](https://github.com/user-attachments/assets/d7f95441-245c-4077-9695-5deb1840cb58)

**Comment by shawnborton** [comment]
Okay nice, I think that can work. What do you think?

**Comment by VickyStash** [comment]
> Okay nice, I think that can work. What do you think?

For me it looks okay as well. Also, that's doesn't look like super often case (sending the message with 20+ emojis only)

**Comment by JmillsExpensify** [comment]
Discussion on-going (Commenting to keep MelvinBot happy).

**Comment by VickyStash** [comment]
**Updates:**
- I've implemented scenario 5 and 6
- Tomorrow, I'll be OOO (15th Aug), but I think on Friday, I'll start the PR preparing/active testing on all of the platforms.

**Comment by VickyStash** [comment]
**Updates:** I've started to prepare the [Draft PR](https://github.com/Expensify/App/pull/47547) and testing over platforms

**Comment by VickyStash** [comment]
**Updates:** 
Continue to work on the Draft PR. 
Today, I have tested the **WEB** platform. I've resolved the remaining TODO related to the correct comment text selection. Only one bug with composer height jumping/increasing when an emoji is entered is left on the WEB. I'll share my suggestions for its solutions tomorrow.

**Comment by VickyStash** [comment]
**Updates:**
I've resolved all of the issues I've found for now and started to combine the PR. I still need to add screenshots and clean the code in several places. If everything goes as expected, I should be able to open the PR for review tomorrow.

**Comment by VickyStash** [comment]
The [PR](https://github.com/Expensify/App/pull/47547) has been opened for the review!

**Comment by VickyStash** [comment]
The [latest updates](https://github.com/Expensify/App/pull/47547#issuecomment-2349256211) can be found in the PR.

Note: I'm going to be OOO next week (Sept 16-20) 🌴

**Comment by tomekzaw** [comment]
FYI Please bump `react-native-live-markdown` to 0.1.149 to include https://github.com/Expensify/react-native-live-markdown/pull/491 that fixes "Unsupported type: inline-image" error on Android.

**Comment by VickyStash** [comment]
**Updates:** We are waiting for [this PR](https://github.com/Expensify/App/pull/49250) with `react-native-live-markdown` update to be merged so we can start another round of the review.

**Comment by VickyStash** [comment]
**Updates:** The PR with `react-native-live-markdown` upgrade was [reverted](https://github.com/Expensify/App/pull/50025). But the new one has already been opened: https://github.com/Expensify/App/pull/50047, so waiting for it. It's not possible to bumb the lib version outside of that PR cause inline images functionality was added in version `0.1.145`.

**Comment by roryabraham** [comment]
@francoisl choosing you at random to take this over while I'm on parental leave. Hope you don't mind 🙃

**Comment by VickyStash** [comment]
**Updates:** the [PR](https://github.com/Expensify/App/pull/47547) is waiting for re-review.

**Comment by JmillsExpensify** [comment]
PRs in progress

**Comment by VickyStash** [comment]
**Updates:** 
- Waiting on the confirmation if the last reported bug is truly a bug and should be fixed (see the [comment](https://github.com/Expensify/App/pull/47547#issuecomment-2476161276))

**Comment by VickyStash** [comment]
**Updates:**
- Resolved conflicts and retested the PR, waiting for another round of review.

**Comment by VickyStash** [comment]
**Updates:**
The [PR](https://github.com/Expensify/App/pull/47547) is waiting for the final review by @francoisl

**Comment by JmillsExpensify** [comment]
Payment summary: $1,000 for @parasharrajat via NewDot. Please make sure that you suggest a regression test before requesting payment. Thanks!

**Comment by JmillsExpensify** [comment]
Regression test created

**Comment by garrettmknight** [comment]
**Payment summary:** $1,000 for @parasharrajat via NewDot.

**Comment by JmillsExpensify** [comment]
$1,000 approved for @parasharrajat
